diff -wu a/dmenu.c b/dmenu.c
--- a/dmenu.c	2015-11-08 23:42:21.000000000 +0100
+++ b/dmenu.c	2015-11-10 18:13:43.937450512 +0100
@@ -141,7 +141,6 @@
 
 	if (lines > 0) {
 		/* draw vertical list */
-		w = mw - x;
 		for (item = curr; item != next; item = item->right) {
 			y += h;
 			if (item == sel)
@@ -151,7 +150,7 @@
 			else
 				drw_setscheme(drw, &scheme[SchemeNorm]);
 
-			drw_text(drw, x, y, w, bh, item->text, 0);
+			drw_text(drw, 0, y, mw, bh, item->text, 0);
 		}
 	} else if (matches) {
 		/* draw horizontal list */
